compiler/quilt/tools/data_transfer.py

```diff
@@ -0,0 +1,282 @@
+"""
+High-level helper functions for uploading and downloading fragments.
+"""
+
+from __future__ import print_function
+import gzip
+import os
+import re
+from shutil import copyfileobj, move
+import tempfile
+from threading import Thread, Lock
+
+import requests
+from requests.adapters import HTTPAdapter
+from requests.packages.urllib3.util.retry import Retry
+from six import iteritems, itervalues
+from tqdm import tqdm
+
+from .hashing import digest_file
+from .util import FileWithReadProgress
+
+
+PARALLEL_UPLOADS = 20
+PARALLEL_DOWNLOADS = 20
+
+S3_CONNECT_TIMEOUT = 30
+S3_READ_TIMEOUT = 30
+S3_TIMEOUT_RETRIES = 3
+
+CONTENT_RANGE_RE = re.compile(r'^bytes (\d+)-(\d+)/(\d+)$')
+
+CHUNK_SIZE = 4096
+
+ZLIB_LEVEL = 2
+
+
+# pyOpenSSL and S3 don't play well together. pyOpenSSL is completely optional, but gets enabled by requests.
+# So... We disable it. That's what boto does.
+# https://github.com/boto/botocore/issues/760
+# https://github.com/boto/botocore/pull/803
+try:
+    from urllib3.contrib import pyopenssl
+    pyopenssl.extract_from_urllib3()
+except ImportError:
+    pass
+
+
+def create_s3_session():
+    """
+    Creates a session with automatic retries on 5xx errors.
+    """
+    sess = requests.Session()
+    retries = Retry(total=3,
+                    backoff_factor=.5,
+                    status_forcelist=[500, 502, 503, 504])
+    sess.mount('https://', HTTPAdapter(max_retries=retries))
+    return sess
+
+def download_fragments(store, obj_urls, obj_sizes):
+    assert len(obj_urls) == len(obj_sizes)
+
+    if not obj_urls:
```

Comment:
This check or at least printing the message should probably move back to command.py (download_fragments doesn't really know why obj_urls is empty if it is).

Reply:
Sure. (I put it there because it's gonna come up again when I add the "materialize" function... but whatever, not a big deal.)